api/service/service.go

```diff
@@ -1,126 +1,111 @@
-	"encoding/json"
+	"github.com/mitchellh/mapstructure"
+	"github.com/spaceuptech/space-api-go/api/model"
+	"github.com/spaceuptech/space-api-go/api/transport/websocket"
 	"log"
+	"sync"
 
 	"github.com/spaceuptech/space-api-go/api/config"
-	"github.com/spaceuptech/space-api-go/api/model"
-	"github.com/spaceuptech/space-api-go/api/transport"
-	"github.com/spaceuptech/space-api-go/api/utils"
 )
 
-type ServiceFunction func(*model.Message, *model.Message, transport.CallBackFunction)
-
-// Service contains the values for the service instance
 type Service struct {
-	config  *config.Config
-	service string
-	id      string
-	funcs   map[string]ServiceFunction
-	// options interface{}
-	// store   interface{}
+	options  *config.Config
+	service  string
+	client   *websocket.Socket
+	function map[string]function
+	mux      sync.RWMutex
 }
 
-// RealtimeRequest is the object sent for realtime requests
-type Payload struct {
-	Auth     []byte `json:"auth"`
-	Token    string `json:"token"`
-	Project  string `json:"project"`
-	Type     string `json:"type"`
-	Id       string `json:"id"`
-	Service  string `json:"service"`
-	Params   []byte `json:"params"`
-	Function string `json:"function"`
-	Error    string `json:"error"`
+type function func(params interface{}, auth map[string]interface{}) (interface{}, error)
+
+func Init(options *config.Config, serviceName string, client *websocket.	Socket) (*Service, error) {
+	client.RegisterOnReconnectCallback(func() {
+		v, err := client.Request(websocket.ServiceRegister, model.ServiceRegisterRequest{Service: serviceName, Project: options.Project, Token: options.Token})
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		responseData, ok := v.(map[string]bool)
+		if !ok {
+			log.Println("service initialization wrong type found")
+			return
+		}
+		data, ok := responseData["ack"]
+		if ok {
+			log.Println("service initialization didn't received acknowledgement from knowledge")
+			return
+		}
+		if !data {
+			log.Println("Could not connect to service")
+			return
+		}
+		log.Println("Service started successfully")
+	})
+
+	return &Service{options: options, service: serviceName, client: client, function: map[string]function{}}, nil
 }
 
-func New(config *config.Config, service string) *Service {
-	return nil
+func (s *Service) RegisterFunc(functionName string, fn function) {
+	s.function[functionName] = fn
 }
 
-// func Init(config *config.Config, service string) *Service {
-// 	id := uuid.NewV1().String()
-// 	var w transport.WebsocketConnection
-// 	var cb transport.CallBackFunctions
-// 	w.RegisterOnReconnectCallback(cb("service-register", {} ))
+func (s *Service) serviceRequest(data interface{}) {
+	reqMap, ok := data.(map[string]interface{})
+	if !ok {
+		log.Println("service request type not found")
+	}
 
-// 	return &Service{config, service, id, make(map[string]ServiceFunction)}
-// }
+	req := model.FunctionsPayload{}
+	if err := mapstructure.Decode(reqMap, &req); err != nil {
+		log.Println(err)
```

Comment:
you should print the error with an appropriate caption. Also, you forgot to send the error and then return.

Reply:
done